File: holy_beacon_sdk/ios/Classes/HolyBeaconSdkPlugin.swift
import Flutter
import UIKit
import CoreBluetooth
import CoreLocation

public class HolyBeaconSdkPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    
    private var centralManager: CBCentralManager?
    private var locationManager: CLLocationManager?
    private var isScanning = false
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "holy_beacon_sdk", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "holy_beacon_sdk/events", binaryMessenger: registrar.messenger())
        let instance = HolyBeaconSdkPlugin()
        
        instance.channel = channel
        instance.eventChannel = eventChannel
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(result: result)
        case "isBluetoothEnabled":
            isBluetoothEnabled(result: result)
        case "requestBluetoothEnable":
            requestBluetoothEnable(result: result)
        case "startScan":
            startScan(result: result)
        case "stopScan":
            stopScan(result: result)
        case "setScanParameters":
            setScanParameters(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(result: @escaping FlutterResult) {
        setupManagers()
        result(nil)
    }
    
    private func setupManagers() {
        centralManager = CBCentralManager(delegate: self, queue: nil)
        locationManager = CLLocationManager()
        locationManager?.delegate = self
    }
    
    private func isBluetoothEnabled(result: @escaping FlutterResult) {
        guard let centralManager = centralManager else {
            result(false)
            return
        }
        result(centralManager.state == .poweredOn)
    }
    
    private func requestBluetoothEnable(result: @escaping FlutterResult) {
        // On iOS, we cannot programmatically enable Bluetooth
        // We can only check the current state
        result(centralManager?.state == .poweredOn)
    }
    
    private func startScan(result: @escaping FlutterResult) {
        guard let centralManager = centralManager else {
            result(FlutterError(code: "BLUETOOTH_NOT_INITIALIZED", 
                              message: "Bluetooth manager not initialized", 
                              details: nil))
            return
        }
        
        guard centralManager.state == .poweredOn else {
            result(FlutterError(code: "BLUETOOTH_DISABLED", 
                              message: "Bluetooth is not powered on", 
                              details: nil))
            return
        }
        
        guard !isScanning else {
            result("ALREADY_SCANNING")
            return
        }
        
        // Request location permission if needed
        if locationManager?.authorizationStatus == .notDetermined {
            locationManager?.requestWhenInUseAuthorization()
        }
        
        // Start scanning for all peripherals
        centralManager.scanForPeripherals(withServices: nil, 
                                        options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
        isScanning = true
        result("SCAN_STARTED")
    }
    
    private func stopScan(result: @escaping FlutterResult) {
        guard isScanning else {
            result("NOT_SCANNING")
            return
        }
        
        centralManager?.stopScan()
        isScanning = false
        result("SCAN_STOPPED")
    }
    
    private func setScanParameters(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // iOS doesn't have as many scan parameter options as Android
        // This method is here for consistency
        result(nil)
    }
    
    private func handlePeripheral(_ peripheral: CBPeripheral, 
                                advertisementData: [String: Any], 
                                rssi: NSNumber) {
        var deviceData: [String: Any] = [:]
        
        deviceData["deviceId"] = peripheral.identifier.uuidString
        deviceData["name"] = peripheral.name ?? ""
        deviceData["rssi"] = rssi.intValue
        deviceData["lastSeen"] = Int(Date().timeIntervalSince1970 * 1000)
        
        // Check for iBeacon data in advertisement
        if let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data {
            if let beaconData = parseIBeacon(data: manufacturerData, deviceName: peripheral.name ?? "") {
                deviceData.merge(beaconData) { _, new in new }
            }
        }
        
        // Set defaults if not parsed as beacon
        if deviceData["uuid"] == nil {
            deviceData["uuid"] = ""
            deviceData["major"] = 0
            deviceData["minor"] = 0
            deviceData["protocol"] = "bleDevice"
            deviceData["verified"] = isHolyDevice(deviceName: peripheral.name ?? "")
        }
        
        eventSink?(deviceData)
    }
    
    private func parseIBeacon(data: Data, deviceName: String) -> [String: Any]? {
        guard data.count >= 25 else { return nil }
        
        let bytes = [UInt8](data)
        
        // Check for Apple Company ID (0x004C) and iBeacon format
        guard bytes[0] == 0x4C && bytes[1] == 0x00 && 
              bytes[2] == 0x02 && bytes[3] == 0x15 else { return nil }
        
        // Extract UUID (16 bytes)
        let uuidBytes = Array(bytes[4..<20])
        let uuid = formatUuid(bytes: uuidBytes)
        
        // Extract Major (2 bytes)
        let major = (Int(bytes[20]) << 8) | Int(bytes[21])
        
        // Extract Minor (2 bytes)  
        let minor = (Int(bytes[22]) << 8) | Int(bytes[23])
        
        // Extract TX Power (1 byte, signed)
        let txPower = Int8(bitPattern: bytes[24])
        
        return [
            "uuid": uuid,
            "major": major,
            "minor": minor,
            "protocol": "ibeacon",
            "verified": true
        ]
    }
    
    private func formatUuid(bytes: [UInt8]) -> String {
        let hex = bytes.map { String(format: "%02X", $0) }.joined()
        return "\(hex.prefix(8))-\(hex.dropFirst(8).prefix(4))-\(hex.dropFirst(12).prefix(4))-\(hex.dropFirst(16).prefix(4))-\(hex.dropFirst(20))"
    }
    
    private func isHolyDevice(deviceName: String) -> Bool {
        let name = deviceName.lowercased()
        return name.contains("holy") || name.contains("kronos") || name.contains("blaze")
    }
}

// MARK: - CBCentralManagerDelegate
extension HolyBeaconSdkPlugin: CBCentralManagerDelegate {
    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Handle Bluetooth state changes
        switch central.state {
        case .poweredOn:
            break
        case .poweredOff:
            if isScanning {
                isScanning = false
                eventSink?(FlutterError(code: "BLUETOOTH_POWERED_OFF",
                                      message: "Bluetooth was powered off during scanning",
                                      details: nil))
            }
        default:
            break
        }
    }
    
    public func centralManager(_ central: CBCentralManager, 
                             didDiscover peripheral: CBPeripheral, 
                             advertisementData: [String : Any], 
                             rssi RSSI: NSNumber) {
        handlePeripheral(peripheral, advertisementData: advertisementData, rssi: RSSI)
    }
}

// MARK: - CLLocationManagerDelegate
extension HolyBeaconSdkPlugin: CLLocationManagerDelegate {
    public func locationManager(_ manager: CLLocationManager, 
                              didChangeAuthorization status: CLAuthorizationStatus) {
        // Handle location permission changes
    }
}

// MARK: - FlutterStreamHandler
extension HolyBeaconSdkPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, 
                        eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: holy_beacon_sdk/android/src/main/kotlin/com/holybeacon/sdk/HolyBeaconSdkPlugin.kt
package com.holybeacon.sdk

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import androidx.core.app.ActivityCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class HolyBeaconSdkPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel : MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothLeScanner: BluetoothLeScanner? = null
    private var isScanning = false
    private val handler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    companion object {
        private const val TAG = "HolyBeaconSdkPlugin"
        private const val METHOD_CHANNEL = "holy_beacon_sdk"
        private const val EVENT_CHANNEL = "holy_beacon_sdk/events"
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL)
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
        
        initializeBluetooth()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> initialize(result)
            "isBluetoothEnabled" -> isBluetoothEnabled(result)
            "requestBluetoothEnable" -> requestBluetoothEnable(result)
            "startScan" -> startScan(result)
            "stopScan" -> stopScan(result)
            "setScanParameters" -> setScanParameters(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun initializeBluetooth() {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothAdapter = bluetoothManager.adapter
        bluetoothLeScanner = bluetoothAdapter?.bluetoothLeScanner
    }

    private fun initialize(result: Result) {
        try {
            if (bluetoothAdapter == null) {
                result.error("BLUETOOTH_NOT_SUPPORTED", "Bluetooth is not supported on this device", null)
                return
            }
            result.success(null)
        } catch (e: Exception) {
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun isBluetoothEnabled(result: Result) {
        result.success(bluetoothAdapter?.isEnabled ?: false)
    }

    private fun requestBluetoothEnable(result: Result) {
        // Note: In modern Android, you cannot programmatically enable Bluetooth
        // This would typically show an intent to ask the user to enable Bluetooth
        result.success(bluetoothAdapter?.isEnabled ?: false)
    }

    private fun startScan(result: Result) {
        if (bluetoothAdapter == null || !bluetoothAdapter!!.isEnabled) {
            result.error("BLUETOOTH_DISABLED", "Bluetooth is not enabled", null)
            return
        }

        if (!hasRequiredPermissions()) {
            result.error("PERMISSIONS_DENIED", "Required permissions not granted", null)
            return
        }

        if (isScanning) {
            result.success("ALREADY_SCANNING")
            return
        }

        try {
            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES)
                .setReportDelay(0)
                .build()

            bluetoothLeScanner?.startScan(null, settings, scanCallback)
            isScanning = true
            result.success("SCAN_STARTED")
        } catch (e: SecurityException) {
            result.error("PERMISSIONS_DENIED", "Security exception: ${e.message}", null)
        } catch (e: Exception) {
            result.error("SCAN_ERROR", "Failed to start scan: ${e.message}", null)
        }
    }

    private fun stopScan(result: Result) {
        if (!isScanning) {
            result.success("NOT_SCANNING")
            return
        }

        try {
            bluetoothLeScanner?.stopScan(scanCallback)
            isScanning = false
            result.success("SCAN_STOPPED")
        } catch (e: SecurityException) {
            result.error("PERMISSIONS_DENIED", "Security exception: ${e.message}", null)
        } catch (e: Exception) {
            result.error("SCAN_ERROR", "Failed to stop scan: ${e.message}", null)
        }
    }

    private fun setScanParameters(call: MethodCall, result: Result) {
        // This method can be used to configure scan parameters
        // For now, just return success
        result.success(null)
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            handleScanResult(result)
        }

        override fun onBatchScanResults(results: MutableList<ScanResult>?) {
            results?.forEach { result ->
                handleScanResult(result)
            }
        }

        override fun onScanFailed(errorCode: Int) {
            val errorMessage = when (errorCode) {
                SCAN_FAILED_ALREADY_STARTED -> "Scan already started"
                SCAN_FAILED_APPLICATION_REGISTRATION_FAILED -> "Application registration failed"
                SCAN_FAILED_FEATURE_UNSUPPORTED -> "Feature unsupported"
                SCAN_FAILED_INTERNAL_ERROR -> "Internal error"
                else -> "Unknown error: $errorCode"
            }
            eventSink?.error("SCAN_FAILED", errorMessage, errorCode)
        }
    }

    private fun handleScanResult(scanResult: ScanResult) {
        try {
            val device = scanResult.device
            val rssi = scanResult.rssi
            val scanRecord = scanResult.scanRecord
            
            val deviceData = mutableMapOf<String, Any>()
            deviceData["deviceId"] = device.address
            deviceData["rssi"] = rssi
            deviceData["lastSeen"] = System.currentTimeMillis()
            
            // Get device name
            val deviceName = try {
                device.name ?: scanRecord?.deviceName ?: ""
            } catch (e: SecurityException) {
                ""
            }
            deviceData["name"] = deviceName
            
            // Parse manufacturer data for iBeacon
            scanRecord?.let { record ->
                val manufacturerData = record.getManufacturerSpecificData(0x004C) // Apple Company ID
                if (manufacturerData != null && manufacturerData.size >= 23) {
                    val beacon = parseIBeacon(manufacturerData, deviceName)
                    if (beacon != null) {
                        deviceData.putAll(beacon)
                    }
                }
            }
            
            // Set default values if not parsed as beacon
            if (!deviceData.containsKey("uuid")) {
                deviceData["uuid"] = ""
                deviceData["major"] = 0
                deviceData["minor"] = 0
                deviceData["protocol"] = "bleDevice"
                deviceData["verified"] = isHolyDevice(deviceName)
            }
            
            eventSink?.success(deviceData)
            
        } catch (e: Exception) {
            // Log error but don't crash
            android.util.Log.e(TAG, "Error processing scan result", e)
        }
    }

    private fun parseIBeacon(data: ByteArray, deviceName: String): Map<String, Any>? {
        if (data.size < 23) return null
        
        try {
            // Check iBeacon format: 0x02, 0x15
            if (data[0] != 0x02.toByte() || data[1] != 0x15.toByte()) return null
            
            // Extract UUID (16 bytes)
            val uuidBytes = data.sliceArray(2..17)
            val uuid = formatUuid(uuidBytes)
            
            // Extract Major (2 bytes)
            val major = ((data[18].toInt() and 0xFF) shl 8) or (data[19].toInt() and 0xFF)
            
            // Extract Minor (2 bytes)
            val minor = ((data[20].toInt() and 0xFF) shl 8) or (data[21].toInt() and 0xFF)
            
            // Extract TX Power (1 byte, signed)
            val txPower = data[22].toInt()
            
            return mapOf(
                "uuid" to uuid,
                "major" to major,
                "minor" to minor,
                "protocol" to "ibeacon",
                "verified" to true
            )
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Error parsing iBeacon", e)
            return null
        }
    }

    private fun formatUuid(bytes: ByteArray): String {
        val hex = bytes.joinToString("") { "%02X".format(it) }
        return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20, 32)}"
    }

    private fun isHolyDevice(deviceName: String): Boolean {
        val name = deviceName.lowercase()
        return name.contains("holy") || name.contains("kronos") || name.contains("blaze")
    }

    private fun hasRequiredPermissions(): Boolean {
        return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
            // Android 12+
            ActivityCompat.checkSelfPermission(context, android.Manifest.permission.BLUETOOTH_SCAN) == PackageManager.PERMISSION_GRANTED &&
            ActivityCompat.checkSelfPermission(context, android.Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        } else {
            // Android 11 and below
            ActivityCompat.checkSelfPermission(context, android.Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        
        if (isScanning) {
            try {
                bluetoothLeScanner?.stopScan(scanCallback)
            } catch (e: SecurityException) {
                // Ignore
            }
        }
    }
}